Rebase on Amazon Linux 2023 and add FIPS support

jpculp opened the pull request:
**Issue number:**

Closes #73, #77

**Description of changes:**

Rebases on Amazon Linux 2023, replacing compiled `lscpu` and `script` in favor of the binaries vendored by Amazon Linux.

Also enable FIPS on FIPS-enabled hosts (based on access to `/proc/sys/crypto/fips_enabled`)

Also update the SSM agent and version.

**Testing done:**

-AL2023-
Launched latest version of Bottlerocket with new control container set in userdata and ensured basic functionality is still there:

- [x] Access to APIclient
- [x] lscpu detects number of socket
- [x] Enable admin container
- [x] Bottlerocket release file is valid
- [x] Bottlerocket package inventory is valid JSON and read-only

-FIPS-
Launched container on a FIPS-enabled host with and without `--mount type=bind,readonly,src=/proc/sys/crypto/fips_enabled,dst=/proc/sys/crypto/fips_enabled`.


**Terms of contribution:**

By submitting this pull request, I agree that this contribution is dual-licensed under the terms of both the Apache License, version 2.0, and the MIT license.

jpculp committed:
Rebase on Amazon Linux 2023

Also replaces compiled `lscpu` and `script` in favor of the binaries
vendored by Amazon Linux.

Signed-off-by: Patrick J.P. Culp <jpculp@amazon.com>

jpculp committed:
Enable FIPS on FIPS-enabled hosts

Signed-off-by: Patrick J.P. Culp <jpculp@amazon.com>

arnaldo2792 commented on start_control_ssm.sh:
```diff
@@ -65,6 +68,17 @@ fetch_from_json() {
 
+if [[ ${FIPS_MODE_FLAG} -eq 1 ]]; then
+  update-crypto-policies --set FIPS 2>/dev/null
+  if [[ "$(cat '/etc/crypto-policies/config')" != "FIPS" ]]; then
+    log "Failed to validate FIPS configuration"
+    exit 1
+  fi
```
Nice!

arnaldo2792 commented on Dockerfile:
```diff
@@ -68,42 +25,46 @@ RUN : \
 
 LABEL "org.opencontainers.image.version"="$IMAGE_VERSION"
 
+# Install the arch specific build of SSM agent *and confirm that it installed* -
+# dnf will allow architecture-mismatched packages to not install and consider
+# the run successful.
```
What's the line that confirms that the agent was installed?

jpculp replied:
This is a really old comment, but I think it's referring to the `check-ssm-agent` Makefile target.

arnaldo2792 replied:
If this is stale, lets clean it up before you merge.

jpculp replied:
Rather than remove it, I just made it valid by adding an extra check.

arnaldo2792 commented on Dockerfile:
```diff
@@ -121,14 +82,14 @@ ENV ENV=/etc/skel/.bashrc
-  ./enable-admin-container \
-  ./enter-admin-container \
-  /usr/bin/
+    ./disable-admin-container \
+    ./enable-admin-container \
+    ./enter-admin-container \
+    /usr/bin/
```
Are these to fix Docker linting problems?

jpculp replied:
Right.

jpculp committed:
Add amazon-ssm-agent check to Dockerfile

Validate that the agent is present when building directly rather than just with
the Makefile.

Signed-off-by: Patrick J.P. Culp <jpculp@amazon.com>

jpculp commented:
- Add additional check for `amazon-ssm-agent`.
- Jump to version **v0.10.0**.

jpculp commented:
Rebased.

jpculp commented:
Add changelog.

jpculp commented:
Jump to v0.20.0 instead to match bottlerocket-admin-container for the AL2023 rebase.

jpculp committed:
Update SSM agent to version 3.3.3572.0

Signed-off-by: Patrick J.P. Culp <jpculp@amazon.com>

jpculp committed:
Version v0.20.0

Signed-off-by: Patrick J.P. Culp <jpculp@amazon.com>

jpculp commented:
Bump SSM agent to latest.
